moneriote.py

```diff
@@ -1,303 +1,557 @@
+        for i, match in enumerate(matches):
+            if match.group(1) == 'white':
+                address = match.group(3)
+                nodes.append(RpcNode(address=address))
+
+        log_msg('Got peers from RPC: %d node(s)' % len(nodes))
+        return nodes
+
+    def scan(self, nodes: RpcNodeList, remove_invalid=False):
+        """
+        Start threads checking known nodes to see if they're alive.
+        :param nodes:
+        :param remove_invalid: only return valid nodes when set to True
+        :return: valid nodes
+        """
+        if len(nodes) == 0:
+            return nodes
+
+        now = datetime.now()
+        log_msg('Scanning %d node(s) on port %d. This can take several minutes. Let it run.' % (
+            len(nodes), self._m_rpc_port))
+
+        pool = Pool(processes=self._concurrent_scans)
+        nodes = RpcNodeList.from_list(pool.map(partial(RpcNode.is_valid, self._blockchain_height), nodes))
+        pool.close()
+        pool.join()
+
+        log_msg('Scanning %d node(s) done after %d seconds, found %d valid' % (
+            len(nodes), (datetime.now() - now).total_seconds(), len(nodes.valid(valid=True))))
+
+        if remove_invalid:
+            nodes = nodes.valid(valid=True)
+        return nodes
+
+    def load_nodes(self):
+        # Verify cached nodes
+        nodes = RpcNodeList.cache_read(PATH_CACHE)
+        nodes = self.scan(nodes, remove_invalid=True)
+
+        # We have enough valid nodes to fill Cloudflare
+        if len(nodes) >= self.cf_max_records:
+            nodes.cache_write()
+            return nodes
+
+        # Ask monerod for more peers and verify
+        nodes += self.scan(self.monerod_get_peers(), remove_invalid=True)
+        nodes.cache_write()
+        return nodes
+
+    def loop(self):
+        # get & set the current blockheight
+        height = self.monerod_get_height(method=self.md_height_discovery_method)
+        if not height or not isinstance(height, int):
+            log_err("Unable to fetch the current blockchain height")
+            return
+
+        self._blockchain_height = height
+
+        # Get the correct zone
+        if not self.cf_zone_id:
+            log_msg('Determining zone_id; looking for \'%s\'' % self.cf_domain_name)
+            zones = self._cf_request(url='')
+            try:
+                self.cf_zone_id = next(zone.get('id') for zone in zones if zone.get('name') == self.cf_domain_name)
+            except StopIteration:
+                log_err('could not determine zone_id. Is your Cloudflare domain correct?')
+                sys.exit()
+            log_msg('Cloudflare zone_id \'%s\' matched to \'%s\'' % (self.cf_zone_id, self.cf_domain_name))
+
+        # Fetch existing A records from Cloudflare
+        nodes = RpcNodeList()
+
+        log_msg('Fetching existing record(s) (%s.%s)' % (self.cf_subdomain_name, self.cf_domain_name))
+        for record in self._cf_request('%s/dns_records/' % self.cf_zone_id):
```

Comment:
I got an exception here.
```
[2018-09-05 20:03] Error contacting Cloudflare (xxxxxxxxxxxxxxxxxxxxx705e181xxxe/dns_records/): HTTPSConnectionPoolhost='api.cloudflare.com', port=443): Read timed out. (read timeout=5)
Traceback (most recent call last):
  File "C:\Users\walass\Desktop\opennode\testing\moneriote.py", line 500, in <module>
    mon.loop()
  File "C:\Users\walass\Desktop\opennode\testing\moneriote.py", line 423, in loop
    for record in self._cf_request('%s/dns_records/' % self.cf_zone_id):
TypeError: 'NoneType' object is not iterable
```

Reply:
Latest commit breaks things. If you were not using my latest commit, then it's a bug and I will fix it later.